types/yup/index.d.ts

```diff
@@ -55,29 +73,43 @@ export interface Schema<T> {
     oneOf(arrayOfValues: any[], message?: string): this;
     notOneOf(arrayOfValues: any[], message?: string): this;
     when(keys: string | any[], builder: WhenOptions<this>): this;
-    test(name: string, message: string, test: (this: TestContext, value?: any) => boolean | ValidationError | Promise<boolean | ValidationError>, callbackStyleAsync?: boolean): this;
+    test(
+        name: string,
+        message:
+            | string
+            | ((params: object & Partial<TestMessageParams>) => string),
```

Comment:
`test` can now take a function as message and it gets passes some data to use to compose an error message. The documentation around this is not all that defined and tracing through the code, I found properties that I listed in `TestMessageParams` but there could be more so I set the type as a union of `object` and the partial.

Reply:
I know this is an old commit - but `TestMessageParams` is missing the property of the method being used (not sure if this is applicable for all methods)... if I use `min` for example with a function for message creation:

```
.min(2, ({ path, value, originalValue, label, min }) => {
  return `String must be at lease ${min} length`;
})
```

Then `min` does not exist on the `TestMessageParams` interface. However Yup passes this through. I'm relatively new to TS so forgive me if something here is incorrect.